File: backend/python/analytics-processor-service/src/rag_indexer/chunking.py
from typing import List, Tuple


def chunk_text(
    text: str,
    max_chars: int = 1200,
    overlap: int = 80,
    min_boundary: int = 120,
    normalize_whitespace: bool = True,
    trim_space: bool = True,
) -> List[Tuple[str, int, int]]:
    if not text:
        return []

    if normalize_whitespace:
        text = " ".join(text.split())
    if trim_space:
        text = text.strip()
    if not text:
        return []

    segments = []
    start = 0
    n = len(text)
    while start < n:
        end = min(start + max_chars, n)
        cut = _choose_boundary(text, start, end, min_boundary)
        chunk = text[start:cut]
        if trim_space:
            chunk = chunk.strip()
        segments.append((chunk, start, cut))
        if cut == n:
            break
        next_start = cut - overlap
        if next_start <= start:
            next_start = cut
        start = next_start
    return segments


def _choose_boundary(text: str, start: int, end: int, min_boundary: int) -> int:
    if end >= len(text):
        return len(text)
    window = text[start:end]
    for chars in ("\n.?!", " "):
        idx = -1
        for ch in chars:
            idx = max(idx, window.rfind(ch))
        if idx >= 0 and idx + 1 >= min_boundary:
            return start + idx + 1
    return end

File: backend/go/libs/platform-rag/pkg/chunking/chunking.go
package chunking

import "strings"

// Options controls how text is chunked before embedding.
type Options struct {
	MaxChars            int  // target maximum characters per chunk
	Overlap             int  // number of chars to overlap between consecutive chunks
	NormalizeWhitespace bool // collapse repeated whitespace into single spaces
	TrimSpace           bool // trim leading/trailing whitespace in each chunk
	MinCharsForBoundary int  // minimum chars before allowing a boundary fallback
}

// Segment represents a slice of text with its original offsets.
type Segment struct {
	Text  string
	Start int
	End   int
}

const (
	defaultMaxChars            = 1200
	defaultOverlap             = 80
	defaultMinCharsForBoundary = 120
)

// ChunkText splits the input into segments respecting max length and overlap.
func ChunkText(input string, opt Options) []Segment {
	if input == "" {
		return nil
	}

	cfg := normalizeOptions(opt)
	text := input
	if cfg.NormalizeWhitespace {
		text = strings.Join(strings.Fields(text), " ")
	}
	if cfg.TrimSpace {
		text = strings.TrimSpace(text)
	}
	if text == "" {
		return nil
	}

	var segments []Segment
	start := 0
	for start < len(text) {
		end := start + cfg.MaxChars
		if end > len(text) {
			end = len(text)
		}

		// Try to cut at a natural boundary inside the window.
		cut := chooseBoundary(text, start, end, cfg.MinCharsForBoundary)

		chunk := text[start:cut]
		if cfg.TrimSpace {
			chunk = strings.TrimSpace(chunk)
		}
		segments = append(segments, Segment{Text: chunk, Start: start, End: cut})

		if cut == len(text) {
			break
		}

		next := cut - cfg.Overlap
		if next <= start {
			next = cut
		}
		start = next
	}

	return segments
}

func chooseBoundary(text string, start, end, minBoundary int) int {
	if end >= len(text) {
		return len(text)
	}
	window := text[start:end]
	if idx := lastBoundary(window, "\n.?!", minBoundary); idx >= 0 {
		return start + idx
	}
	if idx := lastBoundary(window, " ", minBoundary); idx >= 0 {
		return start + idx
	}
	return end
}

func lastBoundary(window string, chars string, minBoundary int) int {
	idx := strings.LastIndexAny(window, chars)
	if idx >= 0 && idx+1 >= minBoundary {
		return idx + 1
	}
	return -1
}

func normalizeOptions(opt Options) Options {
	if opt.MaxChars <= 0 {
		opt.MaxChars = defaultMaxChars
	}
	if opt.Overlap < 0 {
		opt.Overlap = 0
	}
	if opt.Overlap >= opt.MaxChars {
		opt.Overlap = opt.MaxChars / 4
	}
	if opt.MinCharsForBoundary <= 0 {
		opt.MinCharsForBoundary = defaultMinCharsForBoundary
	}
	return opt
}
